blog/2022-11-21-golang-memory-dump.md

```diff
@@ -0,0 +1,187 @@
+A simple way of doing so is to create a Kubernetes service and reach the pod from another container.
+
+```shell
+# Expose the pprof deployment
+kubectl expose deployment pprof-example-app-go --port=8080
+```
+
+Create a curl pod and look at the data inside our app.
+
+```shell
+kubectl run -i -t curl --image=curlimages/curl:latest /bin/sh
+```
+
+You will be sent directly in to the container and you can run something like:
+
+```shell
+curl http://pprof-example-app-go:8080/debug/pprof/allocs?debug=1
+```
+
+This will show you an output that look something like this:
+
+```shell
+heap profile: 4: 47357952 [5473: 554303632] @ heap/1048576
+1: 46882816 [1: 46882816] @ 0x697585 0x470ce1
+#	0x697584	main.allocMem+0xa4	/home/brancz/src/github.com/polarsignals/pprof-example-app-go/main.go:65
+
+1: 212992 [1: 212992] @ 0x4e3b6e 0x4e3e6c 0x6974c5 0x470ce1
+#	0x4e3b6d	log.(*Logger).Output+0x38d	/usr/local/go/src/log/log.go:180
+#	0x4e3e6b	log.Println+0x6b		/usr/local/go/src/log/log.go:329
+#	0x6974c4	main.calculateFib+0xc4		/home/brancz/src/github.com/polarsignals/pprof-example-app-go/main.go:55
+
+1: 204800 [1: 204800] @ 0x4fb68f 0x4fb256 0x503d3d 0x502c97 0x4f6d5e 0x4f6c92 0x4d22a5 0x4d2625 0x4d70b1 0x4cf3d2 0x4e3e3f 0x6974c5 0x470ce1
+#	0x4fb68e	math/big.nat.make+0x5ee		/usr/local/go/src/math/big/nat.go:69
+#	0x4fb255	math/big.nat.sqr+0x1b5		/usr/local/go/src/math/big/nat.go:595
+```
+
+Sadly this output isn't the easiest to read so why not use pprof and while we are at it why not use ephemeral containers.
+
+## Debug container/ephemeral container
+
+Before doing my research my plan was to write about how we attached an volume so we could
+save data locally inside the ephemeral container and then copy it out of the container on to our client and
+do show some nice flame graphs. But apparently it's not supported to attach volumes, it's not even supported to reach the existing volumes on the pod.
+
+There is an open [issue](https://github.com/kubernetes/kubectl/issues/1071) to solve this but it's not part of the current enhancement [proposal](https://github.com/kubernetes/enhancements/issues/1441) so this is nothing that we will see in the near future.
+
+So instead I will just show how we can debug using pprof from within the container.
+
+Since we exposed the endpoint through a service we could of course do this from another pod as well.
+But in general you should be extremely restrictive of what traffick that can reach your pprof endpoint if you expose it at all.
+
+So finally time to use the `kubectl debug` command.
+
+The debug command is used on a specific pod, in my case it's `pprof-example-app-go-7c4b6d77d-xw52p`.
+Let's attach a standard golang container to our running pod, in this case i choose golang 1.15 to match the pprof tool with the running application.
+
+Kubernetes will attach the container for you and give you a shell.
+
+```shell
+kubectl debug pprof-example-app-go-7c4b6d77d-xw52p --image=golang:1.15-alpine3.14 -i -t -- /bin/sh
```

Comment:
The `kubectl run` and this one has arguments in different order. Since `run`, `exec` and `debug` have more or less identical args, I think it would be helpful to put them in the same order. I prefer this order.

Reply:
Agreed